src/base/DeviceValue.hh

```diff
@@ -0,0 +1,42 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file DeviceValue.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Type-safe wrapper to indicate GPU/device storage.
+ *
+ * This should be used primarily to wrap \c Pointers classes returned from host
+ * utility code (e.g. from a \c Store class).
+ * \code
+   struct FooStore
+   {
+      DeviceValue<FooPointers> device_pointers();
+      DeviceValue<const FooPointers> device_pointers() const;
+   };
+   \endcode
+ * used like
+ * \code
+   __global__ void do_stuff(const FooPointers pointers);
+
+   void launch_kernel(const FooStore& foo)
+   {
+       do_stuff<<<...>>>(foo.device_pointers().value);
+   }
+   \endcode
+ */
+template<class T>
+struct DeviceValue
+{
+    T value;
```

Comment:
We might be able to avoid the double curly braces with
```suggestion
    DeviceValue(T &&input) : value(std::move(input)) {};
    T value;
```
or
```suggestion
  template <typename... ArgsTypes>
  DeviceValue(ArgsTypes... params) : value(params...) {};
   T value;
```
The later will make the compiler error message a bit more indrect when getting the parameter list wrong.